### ISSUE TITLE
java.lang.ArrayIndexOutOfBoundsException in TreeSHAP.java:206 when calculating predictions

### ISSUE THREAD

**Issue by dzlab** [h2oai/sparkling-water#2476]
@mn-mikke I'm having a weird issue when trying to calculate predictions when contribution is enabled, it seems to happen only on DRF models.

Can you tell from the logs what could be the probem?

```
Lost task 0.0 in stage 610.0 (TID 4165, 10.0.12.33, executor 1): org.apache.spark.SparkException: Failed to execute user defined function(H2OMOJOPredictionBinomial$$Lambda$7080/288525999: (struct<;gender:string,SeniorCitizen:int,Partner:string,Dependents:string,tenure:int,PhoneService:string,OnlineBackup:string,DeviceProtection:string,Contract:string,PaymentMethod:string,MonthlyCharges:double,TotalCharges:double>;, double) =>; struct<;label:string,probabilities:struct<;No:double,Yes:double>;,contributions:struct<;SeniorCitizen:float,Partner:float,Dependents:float,tenure:float,OnlineBackup:float,DeviceProtection:float,MonthlyCharges:float,TotalCharges:float,gender:float,PhoneService:float,PaymentMethod:float,Contract:float,BiasTerm:float>;>;)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:729)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.util.collection.ExternalSorter.insertAll(ExternalSorter.scala:192)
	at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:62)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:127)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 120
	at hex.genmodel.algos.tree.TreeSHAP$PathPointer.move(TreeSHAP.java:206)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:123)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:181)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:181)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:181)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:181)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.treeShap(TreeSHAP.java:177)
	at hex.genmodel.algos.tree.TreeSHAP.calculateContributions(TreeSHAP.java:237)
	at hex.genmodel.algos.tree.TreeSHAPEnsemble.calculateContributions(TreeSHAPEnsemble.java:29)
	at hex.genmodel.algos.tree.ContributionsPredictor.calculateContributions(ContributionsPredictor.java:26)
	at hex.genmodel.easy.EasyPredictModelWrapper.predictBinomial(EasyPredictModelWrapper.java:644)
	at ai.h2o.sparkling.ml.models.H2OMOJOPredictionBinomial.$anonfun$getBinomialPredictionUDF$1(H2OMOJOPredictionBinomial.scala:45)
	at ai.h2o.sparkling.ml.models.H2OMOJOPredictionBinomial.$anonfun$getBinomialPredictionUDF$1$adapted(H2OMOJOPredictionBinomial.scala:42)
	... 19 more
```

**Comment by mn-mikke** [comment]
This looks like a problem in the MOJO model itself. Can you share a reproducible piece of code that demonstrates this scenario? Is this problem independent on the training dataset?

**Comment by mn-mikke** [comment]
cc @navdeep-G @michalkurka

**Comment by dzlab** [comment]
The train dataset is a clean one with no nulls or NaNs, so I trained an automl model and then tried to calculate predictions on a test set using every model in the leaderboard, all of them worked fine except the DRF that keep crashing with this error.
I will try to extract my training logic into a separately runnable code.

**Comment by mn-mikke** [comment]
> I will try to extract my training logic into a separately runnable code.
@dzlab any update on this issue?

**Comment by dzlab** [comment]
@mn-mikke it is actually very tricky as we are calling H2O API direct during training (to get more control). I will close this ticket and if I can reproduce will try to provide more details.

**Comment by michalkurka** [comment]
@mn-mikke @dzlab I think I found the source of the issue (or at least a very similar issue)

https://h2oai.atlassian.net/browse/PUBDEV-8195 - I will fix ASAP on H2O side

**Comment by michalkurka** [comment]
Fixed in H2O, to be released in 3.32.1.4

**Comment by mn-mikke** [comment]
Hi @dzlab,
SW `3.32.1.4-1` has been released. Can you confirm that the problem was fixed?

**Comment by mn-mikke** [comment]
Hi @dzlab,
does the problem still persist or can we close the issue?

**Comment by dzlab** [comment]
@mn-mikke I think we can close
